grpclog/component.go

```diff
@@ -0,0 +1,233 @@
+
+// init extracts the component settings from the environment variable.
+func init() {
+	// Pull environment variable data and put in environmentVars and prefixVars
+	v, _ := os.LookupEnv(envName)
+	environmentVars, prefixVars = parseEnvironmentVar(v)
+}
+
+// parseEnvironmentVar parses an environment variable string and pulls the component settings data.
+func parseEnvironmentVar(envVar string) (map[string]*componentData, map[string]*componentData) {
+	envVars := map[string]*componentData{}
+	preVars := map[string]*componentData{}
+	if len(envVar) == 0 {
+		return envVars, preVars
+	}
+	varList := strings.Split(envVar, ",")
+	for _, varPair := range varList {
+		varPairList := strings.Split(varPair, ":")
+		if len(varPairList) != 2 {
+			fmt.Fprintf(os.Stderr, "error: could not parse '%v' value '%v', unrecognized key-value pair '%v'\n", envName, envVar, varPair)
+			continue
+		}
+		if cData, ok := parseVar(varPairList[0], varPairList[1]); ok {
+			if varPrefix, ok := getPrefix(varPairList[0]); ok {
+				preVars[varPrefix] = &cData
+			} else {
+				envVars[varPairList[0]] = &cData
+			}
+		} else {
+			fmt.Fprintf(os.Stderr, "error: could not parse '%v' value '%v', unrecognized value '%v'\n", envName, envVar, varPairList[1])
+		}
+	}
+	return envVars, preVars
+}
+
+// apply applies the parameter componentData to the receiver componentData.
+// Sentinel values in the parameter will not be applied.
+func (c *componentData) apply(applyData *componentData) {
+	if applyData.verbosity != sentinel {
+		c.verbosity = applyData.verbosity
+	}
+	if applyData.level != sentinel {
+		c.level = applyData.level
+	}
+}
+
+// parseVar parses a key-value pair from the environment variable. The resulting
+// componentData will have a sentinel value for the verbosity if it is
+// unspecified in the value. Second return value is false if there was an error
+// parsing.
+func parseVar(key string, value string) (componentData, bool) {
+	result := componentData{key, sentinel, sentinel}
+	value = strings.ToUpper(value)
+	switch {
+	case value == "INFO":
+		result.level = levelInfo
+	case value == "WARNING":
+		result.level = levelWarning
+	case value == "ERROR":
+		result.level = levelError
+	case strings.HasPrefix(value, "INFO_"):
+		result.level = levelInfo
+		vStr := strings.TrimPrefix(value, "INFO_")
+		v, err := strconv.Atoi(vStr)
+		if err != nil {
+			return result, false
+		}
+		result.verbosity = v
+	default:
+		return result, false
+	}
+	return result, true
+}
+
+// getPrefix Gets the prefix if s has a wildcard. If s does not have a wildcard,
+// returns "", false.
+func getPrefix(s string) (string, bool) {
+	if strings.HasSuffix(s, "*") {
+		return strings.TrimSuffix(s, "*"), true
+	}
+	return "", false
+}
+
+func (c *componentData) InfoDepth(depth int, args ...interface{}) {
+	if c.level > levelInfo {
+		return
+	}
+	args = append([]interface{}{"[" + string(c.name) + "]"}, args...)
+	grpclog.InfoDepth(depth, args...)
```

Comment:
All component loggers rely on loggerv2.

Loggerv2 needs the old env variables to be set (if they don't use their own logger)
This means, for users to use new env variables, they need to set both old and new env variables. This doesn't make sense.

Reply:
I see this output from the client for the interop client (didn't see any output from the server): https://pastebin.com/YD9VZwn5

Can you give me a way to reproduce to transport log issue?

Regarding channelz, you are right, because the logging is done through the channelz component.

The component logging methods log through whatever logger is set as the default logger, so to me it makes sense that they are set differently. I can't really set the logger with the component logging package, making it so the logger doesn't need to be set explicitly, since that would be assuming things about the nature of the underlying logger.

Reply:
We need to solve the channelz problem. Asking users to set channelz verbose level to control core logging is confusing.

All transport logs are checking `if logger.V(2)`.
Before, setting `GRPC_GO_LOG_VERBOSITY_LEVEL=99` enables all transport logging.
Now, you need to set `GRPC_GO_LOG_LEVEL="transport:INFO_99"`.
This is a behavior change.

To try, cherry pick https://github.com/menghanl/grpc-go/commit/ac9b1f9c4b5b6fc0c24ad608a1854cf1eaceef81, and run `go run myhello/greeter_server/main.go`, `GRPC_GO_LOG_VERBOSITY_LEVEL=99 GRPC_GO_LOG_SEVERITY_LEVEL=info  go run myhello/greeter_client/main.go`. On master, you see `transport: loopyWriter.run returning. connection error: desc = "transport is closing"`, but not with the this PR.